From 81de48dee20cf7f4182924bd0512c1b0b4d77083 Mon Sep 17 00:00:00 2001
From: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
Date: Tue, 18 Aug 2020 09:15:11 +0700
Subject: [PATCH] core: arm: plat-rzg: Add ECC mode checking for shared memory
 region

When we use ECC setting, the memory and size
of DDR we can use may be changed. So that we must modify them to
compatible with the system(and un-trusted OS)

Signed-off-by: Huynh Thanh Hung <hung.huynh.wz@renesas.com>
---
 core/arch/arm/plat-rzg/conf.mk           |  3 +++
 core/arch/arm/plat-rzg/platform_config.h | 31 ++++++++++++++++++++++++
 core/arch/arm/plat-rzg/sub.mk            |  2 +-
 3 files changed, 35 insertions(+), 1 deletion(-)

diff --git a/core/arch/arm/plat-rzg/conf.mk b/core/arch/arm/plat-rzg/conf.mk
index a9621ffa..a13d3b5f 100644
--- a/core/arch/arm/plat-rzg/conf.mk
+++ b/core/arch/arm/plat-rzg/conf.mk
@@ -48,3 +48,6 @@ endif
 
 CFG_WITH_STACK_CANARIES ?= y
 CFG_CORE_HEAP_SIZE ?= 0x20000
+
+RZG_DRAM_ECC ?= 0
+RZG_ECC_FULL ?= 0
diff --git a/core/arch/arm/plat-rzg/platform_config.h b/core/arch/arm/plat-rzg/platform_config.h
index 7bd99011..4315b918 100644
--- a/core/arch/arm/plat-rzg/platform_config.h
+++ b/core/arch/arm/plat-rzg/platform_config.h
@@ -46,25 +46,56 @@
 
 #if defined(PLATFORM_FLAVOR_ek874)
 #define NSEC_DDR_0_BASE     0x47E00000U
+#if (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)	//Apply ECC Full Single setting
+#define NSEC_DDR_0_SIZE     0x3C200000
+#else //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
 #define NSEC_DDR_0_SIZE     0x78200000
+#endif //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
 #endif
+
 #if defined(PLATFORM_FLAVOR_hihope_rzg2m)
+#if (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 1)	//Apply ECC Full Dual setting
 #define NSEC_DDR_0_BASE     0x47E00000U
 #define NSEC_DDR_0_SIZE     0x78200000
+#elif (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)	//Apply ECC Full Single setting
+#define NSEC_DDR_0_BASE     0x47E00000U
+#define NSEC_DDR_0_SIZE     0x4C200000
 #define NSEC_DDR_1_BASE     0x600000000U
+#define NSEC_DDR_1_SIZE     0x40000000
+#else
+#define NSEC_DDR_0_BASE     0x47E00000U
+#define NSEC_DDR_0_SIZE     0x78200000
+#define NSEC_DDR_1_BASE     0x600000000U
 #define NSEC_DDR_1_SIZE     0x80000000
+#endif //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 1)
 #endif
+
 #if defined(PLATFORM_FLAVOR_hihope_rzg2n)
 #define NSEC_DDR_0_BASE     0x47E00000U
 #define NSEC_DDR_0_SIZE     0x78200000
 #define NSEC_DDR_1_BASE     0x480000000U
+#if (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)	//Apply ECC Full Single setting
+#define NSEC_DDR_1_SIZE     0x14000000
+#else //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
 #define NSEC_DDR_1_SIZE     0x80000000
+#endif //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)
 #endif
+
 #if defined(PLATFORM_FLAVOR_hihope_rzg2h)
+#if (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 1)	//Apply ECC Full Dual setting
+#define NSEC_DDR_0_BASE     0x47E00000U
+#define NSEC_DDR_0_SIZE     0x78200000
+#elif (RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 2)	//Apply ECC Full Single setting
+#define NSEC_DDR_0_BASE     0x47E00000U
+#define NSEC_DDR_0_SIZE     0x4C200000
+#define NSEC_DDR_1_BASE     0x600000000U
+#define NSEC_DDR_1_SIZE     0x40000000
+#else
 #define NSEC_DDR_0_BASE     0x47E00000U
 #define NSEC_DDR_0_SIZE     0x78200000
 #define NSEC_DDR_1_BASE     0x500000000U
 #define NSEC_DDR_1_SIZE     0x80000000
+#endif //(RZG_DRAM_ECC == 1 && RZG_ECC_FULL == 1)
 #endif
 
 /* Full GlobalPlatform test suite requires TEE_SHMEM_SIZE to be at least 2MB */
diff --git a/core/arch/arm/plat-rzg/sub.mk b/core/arch/arm/plat-rzg/sub.mk
index 22ef4ad6..c40bd652 100644
--- a/core/arch/arm/plat-rzg/sub.mk
+++ b/core/arch/arm/plat-rzg/sub.mk
@@ -3,4 +3,4 @@ srcs-y += main.c
 srcs-y += rzg_suspend_to_ram.c
 srcs-y += tee_common_otp.c
 
-cppflags-y += -DLTC_NO_PROTOTYPES
+cppflags-y += -DLTC_NO_PROTOTYPES -DRZG_DRAM_ECC=$(RZG_DRAM_ECC) -DRZG_ECC_FULL=$(RZG_ECC_FULL)
--
2.17.1
